### PR TITLE
XDR integrations - adding terminate commands

### PR DESCRIPTION
**Pull request by maimorag**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-11041. 

## Description
Adding support in 2 new API endpoints as commands

### CODE REVIEW

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 1]
```diff
@@ -1572,6 +1572,72 @@ def remove_user_role(self, user_emails: list[str]) -> dict[str, dict[str, str]]:
             }},
         )
 
+    def terminate_process(self, agent_id, instance_id, process_name, incident_id) -> dict[str, dict[str, str]]:
```
Comment (eyalpalo) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:1575]: Lets add typing here

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 2]
```diff
@@ -1572,6 +1572,72 @@ def remove_user_role(self, user_emails: list[str]) -> dict[str, dict[str, str]]:
...
+            request_data["process_name"] = process_name
+        if incident_id:
+            request_data["incident_id"] = incident_id
+        return self._http_request(
+            method='POST',
+            url_suffix="/endpoints/terminate_process/",
+            json_data={"request_data": request_data},
+        )
+
+    def terminate_causality(self, agent_id: str, causality_id: str, process_name: Optional[str], incident_id: Optional[str]) -> dict[str, dict[str, str]]:
```
Comment (eyalpalo) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:1608]: Those methods are practically the same. Let's think of a smart way to reduce code duplication

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 3]
```diff
@@ -1651,7 +1717,8 @@ def run_polling_command(client: CoreClient,
                         results_function: Callable,
                         polling_field: str,
                         polling_value: List,
-                        stop_polling: bool = False) -> CommandResults:
+                        stop_polling: bool = False,
+                        values_raise_error: List = []) -> CommandResults:
```
Comment (eyalpalo) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:1721]: Please add relevant docstring

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 4]
```diff
@@ -1708,6 +1775,8 @@ def run_polling_command(client: CoreClient,
     result = outputs_result_func.get(polling_field) if isinstance(outputs_result_func, dict) else \
         outputs_result_func[0].get(polling_field)
     cond = result not in polling_value if stop_polling else result in polling_value
+    if values_raise_error and result in values_raise_error:
+        raise DemistoException(f"Command {cmd} didn't succeeded, received status {result}")
```
Comment (eyalpalo) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:1779]:
```suggestion
        raise DemistoException(f"The command {cmd} failed. Received status {result}")
```

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 5]
```diff
@@ -4400,3 +4469,76 @@ def get_incidents_command(client, args):
...
+                                                         instance_id=instance_id,
+                                                         process_name=process_name,
+                                                         incident_id=incident_id)
+        action_id = reply_per_instance_id.get("group_action_id")
+        demisto.debug(f'Action terminate process succeeded with action_id={action_id}')
+        replies.append({"action_id": action_id})
+
+    return CommandResults(
+        readable_output=tableToMarkdown(f'Action terminate process created on instance ids: {", ".join(instance_ids)}', replies),
+        outputs={f'{args.get("integration_context_brand", "CoreApiModule")}.TerminateProcess(val.actionId == obj.actionId)': replies},
```
Comment (eyalpalo) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:4504]:
```suggestion
        outputs={f'{args.get("integration_context_brand", "CoreApiModule")}.TerminateProcess(val.actionId && val.actionId == obj.actionId)': replies},
```

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 6]
```diff
@@ -4400,3 +4469,76 @@ def get_incidents_command(client, args):
...
+        action_id = reply_per_instance_id.get("group_action_id")
+        demisto.debug(f'Action terminate process succeeded with action_id={action_id}')
+        replies.append({"action_id": action_id})
+
+    if not replies:
+        raise DemistoException('Terminate causality failed')
+
+    return CommandResults(
+        readable_output=tableToMarkdown(f'Action terminate causality created on {",".join(causality_ids)}', replies),
+        outputs={f'{args.get("integration_context_brand", "CoreApiModule")}.TerminateCausality(val.actionId == obj.actionId)': replies},
```
Comment (eyalpalo) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:4542]: Same as above

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py [hunk 7]
```diff
@@ -4228,3 +4228,69 @@ def test_request_for_bin_file_via_demisto_call(mocker, allow_bin_response):
...
+        {'name': '/api/webapp/public_api/v1/endpoints/terminate_causality',
+         'status': 200,
+         'data': json.dumps({'group_action_id': 2})}
+    ]
+    )
+
+    result = terminate_causality_command(client=client, args={'agent_id': '1', 'causality_id': [
+                                         'causality_id_1', 'causality_id_2']})
+    assert result.readable_output == '### Action terminate causality created on causality_id_1,causality_id_2\n|action_id|\n|---|\n| 1 |\n| 2 |\n'
+    assert result.raw_response == [{'action_id': 1}, {'action_id': 2}]
```
Comment (eyalpalo) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py:4296]: Lets add a test to test the change you made regarding raise DemistoException(f"The command {cmd} failed. Received status {result}")

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py [hunk 8]
```diff
@@ -467,6 +470,45 @@ def main():  # pragma: no cover
         elif command == 'core-get-incidents':
             return_outputs(*get_incidents_command(client, args))
 
+        elif command == 'core-terminate-process':
+            if not is_demisto_version_ge(version=TERMINATE_BUILD_NUM,
+                                         build_number=TERMINATE_SERVER_VERSION):
+                return 'This command is only available for XSIAM 2.4'
```
Comment (eyalpalo) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py:476]: what happens if we run the command in a lower version than the required one? If it failed with specific error, we might want to catch it in the exception instead checking specific build number

File: Packs/Core/Integrations/CortexCoreIR/README.md [hunk 9]
```diff
@@ -2723,4 +2723,56 @@ Builtin Roles with this permission includes: "Investigator", "Responder", "Privi
...
+| action_id | For polling use. | Optional | 
+| interval_in_seconds | Interval in seconds between each poll. | Optional | 
+| timeout_in_seconds | Polling timeout in seconds. | Optional | 
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| Core.TerminateProcess.action_id | String | The action id. | 
+### core-terminate-causality
```
Comment (eyalpalo) [inline Packs/Core/Integrations/CortexCoreIR/README.md:2753]: please add human readable example and command run example for both

**Review by eyalpalo — CHANGES_REQUESTED** [review]
Lets schedule a demo as well

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py (hunk 8 above)
Comment (maimorag) [reply Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py:476]: there isn't specific error :(

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml [hunk 10]
```diff
@@ -3016,6 +3016,70 @@ script:
     - contextPath: Core.ScriptResult.results.retention_date
       description: Timestamp in which the retrieved files will be deleted from the server.
       type: Date
+  - arguments:
+    - description: The agent id.
+      name: agent_id
+      isArray: false
+      required: true
+    - description: The instance id.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml:3024]:
```suggestion
    - description: The agent ID.
      name: agent_id
      isArray: false
      required: true
    - description: The instance ID.
```

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml [hunk 11]
```diff
@@ -3016,6 +3016,70 @@ script:
...
+      name: instance_id
+      isArray: true
+      required: true
+    - description: The process name.
+      name: process_name
+      isArray: false
+      required: false
+    - description: The incident id.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml:3032]:
```suggestion
    - description: The incident ID.
```

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml [hunk 12]
```diff
@@ -3016,6 +3016,70 @@ script:
...
+      name: incident_id
+      isArray: false
+      required: false
+    - description: For polling use.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml:3036]:
```suggestion
    - description: The action ID. For polling use.
```
Is this a comma-separated list of action IDs. ?

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml [hunk 13]
```diff
@@ -3016,6 +3016,70 @@ script:
...
+      isArray: true
+      name: action_id
+      deprecated: true
+    - description: Interval in seconds between each poll.
+      name: interval_in_seconds
+    - description: Polling timeout in seconds.
+      name: timeout_in_seconds
+    name: core-terminate-process
+    description: Terminate a process by it's instance id. Available only for XSIAM 2.4.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml:3045]:
```suggestion
    description: Terminate a process by its instance ID. Available only for XSIAM 2.4.
```

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml [hunk 14]
```diff
@@ -3016,6 +3016,70 @@ script:
...
+      required: true
+    - description: The causality id.
+      name: causality_id
+      isArray: true
+      required: true
+    - description: The process name.
+      name: process_name
+      isArray: false
+      required: false
+    - description: The incident id.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml:3064]:
```suggestion
      description: The action ID.
      type: String
  - arguments:
    - description: The agent ID
      name: agent_id
      isArray: false
      required: true
    - description: The causality ID.
      name: causality_id
      isArray: true
      required: true
    - description: The process name.
      name: process_name
      isArray: false
      required: false
    - description: The incident ID.
```
Is the causality ID a comma-separated list?

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml [hunk 15]
```diff
@@ -3016,6 +3016,70 @@ script:
...
+      name: incident_id
+      isArray: false
+      required: false
+    - description: For polling use.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml:3068]:
```suggestion
    - description: The action ID. For polling use.
```
Is this a comma-separated list of action IDs?

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml [hunk 16]
```diff
@@ -3016,6 +3016,70 @@ script:
...
+      isArray: true
+      name: action_id
+      deprecated: true
+    - description: Interval in seconds between each poll.
+      name: interval_in_seconds
+    - description: Polling timeout in seconds.
+      name: timeout_in_seconds
+    name: core-terminate-causality
+    description: Terminate a process tree by it's causality id. Available only for XSIAM 2.4.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.yml:3077]:
```suggestion
    description: Terminate a process tree by its causality ID. Available only for XSIAM 2.4.
```

File: Packs/Core/Integrations/CortexCoreIR/README.md [hunk 17]
```diff
@@ -2814,3 +2814,90 @@ Initiates a new endpoint script execution action using a script from the script
 >|---|---|---|---|---|---|---|---|---|---|---|
 >||222|111|1.1.1.1|test|STATUS_010_CONNECTED|COMPLETED_SUCCESSFULLY|0||0||
 
+
+### core-terminate-process
+
+***
+Terminate a process by it's instance id.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/README.md:2821]:
```suggestion
Terminate a process by its instance ID.
```

File: Packs/Core/Integrations/CortexCoreIR/README.md [hunk 18]
```diff
@@ -2814,3 +2814,90 @@ Initiates a new endpoint script execution action using a script from the script
...
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| agent_id | The agent id. | Required | 
+| instance_id | The instance id. | Required | 
+| process_name | The process name. | Optional | 
+| incident_id | The incident id. | Optional | 
+| action_id | For polling use. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/README.md:2835]:
```suggestion
| agent_id | The agent ID. | Required | 
| instance_id | The instance ID. | Required | 
| process_name | The process name. | Optional | 
| incident_id | The incident ID. | Optional | 
| action_id | The action ID. For polling use. | Optional | 
```

File: Packs/Core/Integrations/CortexCoreIR/README.md [hunk 19]
```diff
@@ -2814,3 +2814,90 @@ Initiates a new endpoint script execution action using a script from the script
...
+| incident_id | The incident id. | Optional | 
+| action_id | For polling use. | Optional | 
+| interval_in_seconds | Interval in seconds between each poll. | Optional | 
+| timeout_in_seconds | Polling timeout in seconds. | Optional | 
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| Core.TerminateProcess.action_id | String | The action id. |
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/README.md:2843]:
```suggestion
| Core.TerminateProcess.action_id | String | The action ID. | 
```

File: Packs/Core/Integrations/CortexCoreIR/README.md [hunk 20]
```diff
@@ -2814,3 +2814,90 @@ Initiates a new endpoint script execution action using a script from the script
...
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| Core.TerminateProcess.action_id | String | The action id. | 
+### core-terminate-causality
+
+***
+Stops a process by it's causality id.
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/README.md:2847]:
```suggestion
Stops a process by its causality ID.
```

File: Packs/Core/Integrations/CortexCoreIR/README.md [hunk 21]
```diff
@@ -2814,3 +2814,90 @@ Initiates a new endpoint script execution action using a script from the script
...
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| agent_id | The agent id. | Required | 
+| causality_id | The causality id. | Required | 
+| process_name | The process name. | Optional | 
+| incident_id | The incident id. | Optional | 
+| action_id | For polling use. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/Core/Integrations/CortexCoreIR/README.md:2878]:
```suggestion
| agent_id | The agent ID. | Required | 
| causality_id | The causality ID. | Required | 
| process_name | The process name. | Optional | 
| incident_id | The incident ID. | Optional | 
| action_id | The action ID. For polling use. | Optional | 
```

File: Packs/Core/ReleaseNotes/3_0_69.md [hunk 22]
```diff
@@ -0,0 +1,8 @@
+
+#### Integrations
+
+##### Investigation & Response
+
+- Added the following commands:
+   - **_core-terminate-process_**
+   - **_core-terminate-causality_**
```
Comment (ShirleyDenkberg) [inline Packs/Core/ReleaseNotes/3_0_69.md:8]:
```suggestion
Added the following commands:
 - ***core-terminate-process***
- ***core-terminate-causality***
```

File: Packs/ctf01/ReleaseNotes/1_0_29.md [hunk 23]
```diff
@@ -0,0 +1,7 @@
+<!-- 
+#### Integrations
+
+##### Cortex XDR - IR CTF
+
+No changes related directly to this integration.
```
Comment (ShirleyDenkberg) [inline Packs/ctf01/ReleaseNotes/1_0_29.md:6]: Is this necessary?

File: Packs/CortexXDR/ReleaseNotes/6_1_77.md [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+
+No changes related directly to this integration.
+
+##### Cortex XDR - IOC
+
+No changes related directly to this integration
+
+##### Cortex XDR - XQL Query Engine
+
+No changes related directly to this integration
```
Comment (ShirleyDenkberg) [inline Packs/CortexXDR/ReleaseNotes/6_1_77.md:14]: Is this necessary?

**Comment by ShirleyDenkberg** [comment]
@eyalpalo Doc review completed.

**Review by eyalpalo — APPROVED** [review]
Nice

**Pull request merged**